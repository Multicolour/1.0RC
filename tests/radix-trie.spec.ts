import {
  CreateTrie,
  InsertNodeIntoTrie,
  NodeType,
  SearchTrie,
} from "@lib/server/radix-trie"

type TestData = string

const testTrie = CreateTrie<TestData>()

test("Insert first node", () => {
  InsertNodeIntoTrie<TestData>(testTrie, "/super", "SUPER")
  expect(testTrie).toEqual({
    text: "",
    type: NodeType.PLAIN,
    nodes: [
      {
        text: "/super",
        data: "SUPER",
        type: NodeType.PLAIN,
      },
    ],
  })
})

test("Insert second node", () => {
  InsertNodeIntoTrie<TestData>(testTrie, "/sucky", "SUCKY")

  expect(testTrie).toEqual({
    text: "/su",
    type: NodeType.PLAIN,
    nodes: [
      {
        text: "per",
        data: "SUPER",
        type: NodeType.PLAIN,
      },
      {
        text: "cky",
        data: "SUCKY",
        type: NodeType.PLAIN,
      },
    ],
  })

  InsertNodeIntoTrie<TestData>(testTrie, "/cats", "CATS")

  expect(testTrie).toEqual({
    text: "",
    type: NodeType.PLAIN,
    nodes: [
      {
        text: "/",
        nodes: [
          {
            text: "super",
            data: "SUPER",
            type: NodeType.PLAIN,
          },
          {
            text: "cats",
            data: "CATS",
            type: NodeType.PLAIN,
          },
        ],
      },
    ],
  })
})

// @TODO Use the InsertNodeIntoTrie to create this trie.
const testTrieA = CreateTrie<TestData>("/")
testTrieA.nodes = [
  {
    text: "/",
    nodes: [
      {
        text: "s",
        nodes: [
          {
            text: "uper",
            data: "SUPER",
          },
          {
            text: "illy",
            data: "SILLY",
          },
        ],
      },
      {
        text: "c",
        nodes: [
          {
            text: "ats",
            data: "CATS",
          },
          {
            text: "o",
            nodes: [
              {
                text: "py",
                data: "COPY",
              },
              {
                text: "n",
                nodes: [
                  {
                    text: "tact",
                    data: "CONTACT",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]

test("Basic /cats", () => {
  expect(SearchTrie<TestData>(testTrieA, "/cats")).toEqual({
    text: "ats",
    data: "CATS",
  })
})

test("Basic /contact", () => {
  expect(SearchTrie<TestData>(testTrieA, "/contact")).toEqual({
    text: "tact",
    data: "CONTACT",
  })
})

test("Basic /copy", () => {
  expect(SearchTrie<TestData>(testTrieA, "/copy")).toEqual({
    text: "py",
    data: "COPY",
  })
})

test("Basic /super", () => {
  expect(SearchTrie<TestData>(testTrieA, "/super")).toEqual({
    text: "uper",
    data: "SUPER",
  })
})

test("Basic /silly", () => {
  expect(SearchTrie<TestData>(testTrieA, "/silly")).toEqual({
    text: "illy",
    data: "SILLY",
  })
})
