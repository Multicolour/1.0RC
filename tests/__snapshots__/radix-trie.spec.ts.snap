// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Naughty strings 1`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 2`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 3`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 4`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 5`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 6`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
    Object {
      "data": "😅",
      "nodes": Array [],
      "text": "😅",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 7`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
    Object {
      "data": "😅",
      "nodes": Array [],
      "text": "😅",
    },
    Object {
      "data": "😂",
      "nodes": Array [],
      "text": "😂",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 8`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
    Object {
      "data": "😅",
      "nodes": Array [],
      "text": "😅",
    },
    Object {
      "data": "😂",
      "nodes": Array [],
      "text": "😂",
    },
    Object {
      "data": "🤣",
      "nodes": Array [],
      "text": "🤣",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 9`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
    Object {
      "data": "😅",
      "nodes": Array [],
      "text": "😅",
    },
    Object {
      "data": "😂",
      "nodes": Array [],
      "text": "😂",
    },
    Object {
      "data": "🤣",
      "nodes": Array [],
      "text": "🤣",
    },
    Object {
      "data": "☺️",
      "nodes": Array [],
      "text": "☺️",
    },
  ],
  "text": "",
}
`;

exports[`Naughty strings 10`] = `
Object {
  "nodes": Array [
    Object {
      "data": "😀",
      "nodes": Array [],
      "text": "😀",
    },
    Object {
      "data": "😃",
      "nodes": Array [],
      "text": "😃",
    },
    Object {
      "data": "😄",
      "nodes": Array [],
      "text": "😄",
    },
    Object {
      "data": "😁",
      "nodes": Array [],
      "text": "😁",
    },
    Object {
      "data": "😆",
      "nodes": Array [],
      "text": "😆",
    },
    Object {
      "data": "😅",
      "nodes": Array [],
      "text": "😅",
    },
    Object {
      "data": "😂",
      "nodes": Array [],
      "text": "😂",
    },
    Object {
      "data": "🤣",
      "nodes": Array [],
      "text": "🤣",
    },
    Object {
      "data": "☺️",
      "nodes": Array [],
      "text": "☺️",
    },
    Object {
      "data": "😊",
      "nodes": Array [],
      "text": "😊",
    },
  ],
  "text": "",
}
`;
